jupyterhub/services/auth.py

```diff
@@ -598,15 +598,54 @@ def _get_user_cookie(self, handler):
         token = handler.get_secure_cookie(self.cookie_name)
         session_id = self.get_session_id(handler)
         if token:
+            if self.strict_session_id and not session_id:
+                app_log.warning("Rejecting token in cookie with no session id")
+                handler.clear_cookie(self.cookie_name)
+                return
             token = token.decode('ascii', 'replace')
             user_model = self.user_for_token(token, session_id=session_id)
             if user_model is None:
                 app_log.warning("Token stored in cookie may have expired")
                 handler.clear_cookie(self.cookie_name)
+            # _after_ identifying the user, check the session id
+            elif self.strict_session_id:
+                # user identified, check session id
+                log_user = f"{user_model['kind']}: {user_model['name']}"
+                # check session_id against user model field
+                token_session_id = user_model.get("session_id")
+                if session_id != token_session_id:
+                    app_log.warning(
+                        f"Strict session id mismatch for {log_user} {session_id} != {token_session_id}"
+                    )
+                    handler.clear_cookie(self.cookie_name)
+                    user_model = None
             return user_model
 
     # HubOAuth API
 
+    strict_session_id = Bool(
+        True,
+        help="""Strict session-id checking.
+
+        Requires session-id to match the token in a cookie.
+
+        If a session-id cookie is not set, no cookie-based auth will be accepted.
+        Additionally, the token in the cookie must be associated with the session id to be accepted.
+
+        Can be disabled via JUPYTERHUB_STRICT_SESSION_ID=0 to preserve JupyterHub 1.x behavior.
+
+        .. versionadded:: 2.0
+        """,
+    ).tag(config=True)
```

Comment:
This help text could benefit from a clarification that this is configuration relevant on the singleuser-server rather than the jupyterhub server, if those are separate.

Reply:
Such a note probably goes in the docstring for HubAuth and HubOAuth, to make it clear that these are set in the single-user and/or service environment.